// @maxbcm/webui-sdk/dist/Models/ErrorDialogState.js
export var ErrorDialogStateType;
(function (ErrorDialogStateType) {
    ErrorDialogStateType[ErrorDialogStateType["HIDDEN"] = 0] = "HIDDEN";
    ErrorDialogStateType[ErrorDialogStateType["GENERIC"] = 1] = "GENERIC";
    ErrorDialogStateType[ErrorDialogStateType["WITH_SECONDARY"] = 2] = "WITH_SECONDARY";
    ErrorDialogStateType[ErrorDialogStateType["WITH_BUTTON"] = 3] = "WITH_BUTTON";
    ErrorDialogStateType[ErrorDialogStateType["WITH_BUTTON_AND_SECONDARY"] = 4] = "WITH_BUTTON_AND_SECONDARY";
})(ErrorDialogStateType || (ErrorDialogStateType = {}));
export class ErrorDialogState {
    stateType;
    title;
    message1;
    message2;
    buttonTitle;
    selector;
    constructor(stateType, title, message1, message2, buttonTitle, selector) {
        this.stateType = stateType;
        this.title = title;
        this.message1 = message1;
        this.message2 = message2;
        this.buttonTitle = buttonTitle;
        this.selector = selector;
    }
    static hidden() {
        return new ErrorDialogState(ErrorDialogStateType.HIDDEN, undefined, undefined, undefined, undefined, undefined);
    }
    static generic(title, message) {
        return new ErrorDialogState(ErrorDialogStateType.GENERIC, title, message, undefined, undefined, undefined);
    }
    static withSecondary(title, message1, message2) {
        return new ErrorDialogState(ErrorDialogStateType.WITH_SECONDARY, title, message1, message2, undefined, undefined);
    }
    static withButton(title, message, buttonTitle, selector) {
        return new ErrorDialogState(ErrorDialogStateType.WITH_BUTTON, title, message, undefined, buttonTitle, selector);
    }
    static withButtonAndSecondary(title, message1, message2, buttonTitle, selector) {
        return new ErrorDialogState(ErrorDialogStateType.WITH_BUTTON_AND_SECONDARY, title, message1, message2, buttonTitle, selector);
    }
}
