// @maxbcm/webui-sdk/dist/Models/Server.js
export var ServerCommunicationError;
(function (ServerCommunicationError) {
    ServerCommunicationError[ServerCommunicationError["REQUEST_FAILED"] = 0] = "REQUEST_FAILED";
    ServerCommunicationError[ServerCommunicationError["UNEXPECTED_RESPONSE"] = 1] = "UNEXPECTED_RESPONSE";
    ServerCommunicationError[ServerCommunicationError["UNAUTHORIZED"] = 2] = "UNAUTHORIZED";
    ServerCommunicationError[ServerCommunicationError["DATABASE_PROBLEM"] = 3] = "DATABASE_PROBLEM";
    ServerCommunicationError[ServerCommunicationError["BAD_DB_ENPOINT"] = 4] = "BAD_DB_ENPOINT";
    ServerCommunicationError[ServerCommunicationError["CODED_RESPONSE"] = 5] = "CODED_RESPONSE";
})(ServerCommunicationError || (ServerCommunicationError = {}));
class ServerInfo {
    developmentUrl;
    productionUrl;
    constructor(developmentUrl, productionurl) {
        this.developmentUrl = developmentUrl;
        this.productionUrl = productionurl;
    }
    url(forceProduction = false) {
        let url;
        if (window.location.host.includes("localhost") && !forceProduction) {
            url = this.developmentUrl;
        }
        else {
            url = this.productionUrl;
        }
        return url;
    }
}
export const Server = Object.freeze({
    CENTRAL_HUB: new ServerInfo('http://localhost:11111', 'https://central-hub-c18a96142dc1.herokuapp.com/'),
    NODE: new ServerInfo('http://localhost:5555', 'https://bcm-server-83ddaa46f17a.herokuapp.com'),
    WEBSITE: new ServerInfo('http://localhost:8888', 'https://bcm.software')
});
