// @maxbcm/webui-sdk/dist/Views/ErrorDialog/ErrorDialog.js
import { ErrorDialogStateType } from '../../Models/ErrorDialogState';
import './ErrorDialog.css';
import warningIcon from './Resources/IconWarningWhite.png';
export class ErrorDialog {
    title;
    body;
    innerError;
    actionButton;
    element;
    constructor() {
        const icon = document.createElement('img');
        icon.classList.add('icon');
        icon.src = warningIcon;
        const title = document.createElement('div');
        title.classList.add('title');
        const titleBox = document.createElement('div');
        titleBox.classList.add('titleBox');
        titleBox.appendChild(icon);
        titleBox.appendChild(title);
        const body = document.createElement('div');
        body.classList.add('body');
        const innerError = document.createElement('div');
        innerError.classList.add('innerError');
        const actionButton = document.createElement('button');
        actionButton.classList.add('actionButton');
        actionButton.textContent = 'Retry';
        const errorMessageBox = document.createElement('div');
        errorMessageBox.classList.add('errorMessageBox');
        errorMessageBox.appendChild(titleBox);
        errorMessageBox.appendChild(body);
        errorMessageBox.appendChild(innerError);
        errorMessageBox.appendChild(actionButton);
        this.title = title;
        this.body = body;
        this.actionButton = actionButton;
        this.innerError = innerError;
        this.element = errorMessageBox;
    }
    updateState(state) {
        switch (state.stateType) {
            case ErrorDialogStateType.HIDDEN:
                this.element.style.display = 'none';
                break;
            case ErrorDialogStateType.GENERIC:
                this.element.style.display = 'flex';
                this.title.textContent = state.title;
                this.body.textContent = state.message1;
                this.innerError.style.display = 'none';
                this.actionButton.style.display = 'none';
                break;
            case ErrorDialogStateType.WITH_SECONDARY:
                this.element.style.display = 'flex';
                this.title.textContent = state.title;
                this.body.textContent = state.message1;
                this.innerError.style.display = 'block';
                this.innerError.textContent = state.message2;
                this.actionButton.style.display = 'none';
                break;
            case ErrorDialogStateType.WITH_BUTTON:
                this.element.style.display = 'flex';
                this.title.textContent = state.title;
                this.body.textContent = state.message1;
                this.innerError.style.display = 'none';
                this.actionButton.style.display = 'block';
                this.actionButton.textContent = state.buttonTitle;
                this.actionButton.onclick = state.selector;
                break;
            case ErrorDialogStateType.WITH_BUTTON_AND_SECONDARY:
                this.element.style.display = 'flex';
                this.title.textContent = state.title;
                this.body.textContent = state.message1;
                this.innerError.style.display = 'block';
                this.innerError.textContent = state.message2;
                this.actionButton.style.display = 'block';
                this.actionButton.textContent = state.buttonTitle;
                this.actionButton.onclick = state.selector;
                break;
            default:
                break;
        }
    }
}
